### PR TITLE
remote: verify cache exists on dir checksum

### LINKED ISSUE #2647: remote: getting a directory checksum requires uploading a file to an external cache
**Version**: 0.62.1

**Description**: The current approach to calculate a **directory checksum** is to collect its contents' checksum and then "upload" it and get the checksum provided by the remote:

https://github.com/iterative/dvc/blob/a091b67341b00fea8ef43906f59c88cb38f384ff/dvc/remote/base.py#L246-L256

For example, if it is an S3 remote, DVC will **create a file** containing an "index" of the directory **with its checksums** on the S3 cache **and then will try to get the ETag of such file**.

This **is an easy way to check if the directory has changed**. Since the "directory index" is already there, **you are a `head` operation away** from knowing if the directory changed or not. Otherwise, you'll need to recompute the file on the spot.

Right now **we don't require setting up a cache for using external dependencies, but as mentioned above, this is needed for directories.**

Let's discuss if there's another way around uploading files or just settle with that implementation and modify the code and docs accordingly.

**Related**: #1654

### PR DESCRIPTION
**Pull request by pared**
* [x] ❗ Have you followed the guidelines in the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) list?

* [x] 📖 Check this box if this PR **does not** require [documentation](https://dvc.org/doc) updates, or if it does **and** you have created a separate PR in [dvc.org](https://github.com/iterative/dvc.org) with such updates (or at least opened an issue about it in that repo). Please link below to your PR (or issue) in the [dvc.org](https://github.com/iterative/dvc.org) repo.

* [x] ❌ Have you checked DeepSource, CodeClimate, and other sanity checks below? We consider their findings recommendatory and don't expect everything to be addressed. Please review them carefully and fix those that actually improve code or fix bugs.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Fixes #2647 

[asciicast](https://asciinema.org/a/295538)

### CODE REVIEW

File: dvc/remote/base.py [hunk 1]
```diff
@@ -233,6 +241,9 @@ def _collect_dir(self, path_info):
         return sorted(result, key=itemgetter(self.PARAM_RELPATH))
 
     def get_dir_checksum(self, path_info):
+        if not self.cache:
```
Comment (efiop) [inline dvc/remote/base.py]: Would it make sense to unify it somehow with https://github.com/iterative/dvc/blob/master/dvc/output/base.py#L102 ?

Comment (pared) [reply dvc/remote/base.py]: @efiop Do you mean, using same error in both places, or using `self.REMOTE.scheme` to point user in right direction?

Comment (efiop) [reply dvc/remote/base.py]: @pared Well, maybe both? :)

File: dvc/exceptions.py [hunk 2]
```diff
@@ -310,3 +310,11 @@ def __init__(self, path, repo):
             " neither as an output nor a git-handled file."
         )
         super().__init__(msg.format(path, repo))
+
+
+class RemoteCacheRequiredError(DvcException):
+    def __init__(self, scheme):
+        super().__init__(
+            "Remote cache for '{}' remote is required to perform this "
+            "operation".format(scheme)
```
Comment (efiop) [inline dvc/exceptions.py]: Could we add something helpful or add a link to docs?

Comment (efiop) [reply dvc/exceptions.py]: Also "this operation" is not very clear. Maybe we could include path too to be more precise about what path has caused this exception?

**Pull request merged**